### ISSUE TITLE
device client will never connect if started before networking is ready

### ISSUE THREAD

**Issue by larsonmpdx** [Azure/azure-iot-sdk-c#746]
- **OS and version used:** <VERSION>  buildroot linux 4.1.15 on arm

- **SDK version used:** <VERSION>  Nov 20 2018 1.2.11

# Description of the issue:
If the iot device client application is started before network connectivity is ready (dns not yet working and connections can't be opened), it can get stuck and may never connect even with unlimited time and retries.  if it's started after network connectivity is ready it will work normally and can become disconnected through network faults and then reconnect later and so on, working as expected.

# Code sample exhibiting the issue:
an application more or less the same as https://github.com/Azure/azure-iot-sdk-c/blob/master/iothub_client/samples/iothub_ll_client_x509_sample/iothub_ll_client_x509_sample.c
but running in a loop and relying on the iot library's retry functionality

# Console log of the issue:
this is repeated forever, each time the iot library tries to reconnect (at some interval based on the azure iot reconnect policy):
````
Error: Time:Thu Nov 29 17:04:08 2018 File:/azure-iot-sdk-c/c-utility/adapters/socketio_berkeley.c Func:wait_for_connection Line:386 Failure: connect failure 111.
Error: Time:Thu Nov 29 17:04:08 2018 File:/azure-iot-sdk-c/c-utility/adapters/socketio_berkeley.c Func:socketio_open Line:770 wait_for_connection failed
Error: Time:Thu Nov 29 17:04:08 2018 File:/azure-iot-sdk-c/c-utility/adapters/tlsio_openssl.c Func:on_underlying_io_open_complete Line:760 Invalid tlsio_state. Expected state is TLSIO_STATE_OPENING_UNDERLYING_IO.
Error: Time:Thu Nov 29 17:04:08 2018 File:/azure-iot-sdk-c/c-utility/adapters/tlsio_openssl.c Func:tlsio_openssl_open Line:1258 Failed opening the underlying I/O.
Error: Time:Thu Nov 29 17:04:08 2018 File:/azure-iot-sdk-c/umqtt/src/mqtt_client.c Func:mqtt_client_connect Line:1001 Error: io_open failed
Error: Time:Thu Nov 29 17:04:08 2018 File:/azure-iot-sdk-c/iothub_client/src/iothubtransport_mqtt_common.c Func:SendMqttConnectMsg Line:2141 failure connecting to address [address].
````

**Comment by larsonmpdx** [comment]
I can't reproduce this in a linux VM (by disabling networking or playing with DNS) but it's easy for me to reproduce with my cell-connected device.  Let me know if you want some additional logging or console statements inserted somewhere to track this down

**Comment by larsonmpdx** [comment]
I'm also getting a failure to reconnect when the cell connection drops on our device, with these messages in a loop:
````
Error: Time:Tue Dec  4 00:59:49 2018 File:/azure-iot-sdk-c/c-utility/adapters/socketio_berkeley.c Func:lookup_address_and_initiate_socket_connection Line:282 Failure: getaddrinfo failure -3.
Error: Time:Tue Dec  4 00:59:49 2018 File:/azure-iot-sdk-c/c-utility/adapters/socketio_berkeley.c Func:socketio_open Line:765 lookup_address_and_connect_socket failed
Info: Closing tlsio from a state other than TLSIO_STATE_EXT_OPEN or TLSIO_STATE_EXT_ERROR
Error: Time:Tue Dec  4 00:59:49 2018 File:/azure-iot-sdk-c/c-utility/adapters/tlsio_openssl.c Func:on_underlying_io_open_complete Line:760 Invalid tlsio_state. Expected state is TLSIO_STATE_OPENING_UNDERLYING_IO.
Error: Time:Tue Dec  4 00:59:49 2018 File:/azure-iot-sdk-c/c-utility/adapters/tlsio_openssl.c Func:tlsio_openssl_open Line:1258 Failed opening the underlying I/O.
Error: Time:Tue Dec  4 00:59:49 2018 File:/azure-iot-sdk-c/umqtt/src/mqtt_client.c Func:mqtt_client_connect Line:1000 Error: io_open failed
Error: Time:Tue Dec  4 00:59:49 2018 File:/azure-iot-sdk-c/iothub_client/src/iothubtransport_mqtt_common.c Func:SendMqttConnectMsg Line:2122 failure connecting to address phc-iothub-msg-eastus2-dev.azure-devices.net.
````

I can easily trigger this by setting up the device with a good cell connection and the azure client connected, then running "nmcli radio wwan off" followed by "nmcli radio wwan on".  If I run a ping before or after it succeeds (the cell connection goes away and comes back) but the azure client will never reconnect no matter how long I wait.

**Comment by larsonmpdx** [comment]
see my PR, I can use this demo program on a cell-connected device and start up that device with no network connectivity, then run the demo program and it won't connect.  then while the demo program is still running I can start the network (and pings work) and the demo program will never connect

the default retry policy with the doubling timeouts makes this a little tedious to study after it gets into multiple minute retries but it does reproduce the problem every time for me

**Comment by larsonmpdx** [comment]
it's 2.21

I'm calling res_init() manually in my full program and doesn't help, I'll add it to the demo and see what happens

**Comment by larsonmpdx** [comment]
the res_init() call only fixed my getaddrinfo failure -3, the other error remains

I have a new problem which is this line repeated after a few minutes of being disconnected:

````
Error: Time:Wed Dec  5 18:49:03 2018 File:/azure-iot-sdk-c/iothub_client/src/iothubtransport_mqtt_common.c Func:InitializeConnection Line:2220 mqtt_client timed out waiting for CONNACK
Error: Time:Wed Dec  5 18:49:03 2018 File:/azure-iot-sdk-c/iothub_client/src/iothubtransport_mqtt_common.c Func:InitializeConnection Line:2220 mqtt_client timed out waiting for CONNACK
Error: Time:Wed Dec  5 18:49:03 2018 File:/azure-iot-sdk-c/iothub_client/src/iothubtransport_mqtt_common.c Func:InitializeConnection Line:2220 mqtt_client timed out waiting for CONNACK
Error: Time:Wed Dec  5 18:49:03 2018 File:/azure-iot-sdk-c/iothub_client/src/iothubtransport_mqtt_common.c Func:InitializeConnection Line:2220 mqtt_client timed out waiting for CONNACK
...
````

once this error starts it floods my logs and makes it hard to track down anything else

**Comment by larsonmpdx** [comment]
right before I get the many repeated "CONNACK" error messages in my last comment I get this (timestamps don't match, it's from a different series, but it does happen in this order):
````
[2018-12-05 11:38:02.834] Error: Time:Wed Dec  5 17:37:35 2018 File:/azure-iot-sdk-c/c-utility/adapters/tlsio_openssl.c Func:tlsio_openssl_send Line:1351 Invalid tlsio_state. Expected state is TLSIO_STATE_OPEN.
[2018-12-05 11:38:02.834] Error: Time:Wed Dec  5 17:37:35 2018 File:/azure-iot-sdk-c/umqtt/src/mqtt_client.c Func:sendPacketItem Line:371 1352: Failure sending control packet data
[2018-12-05 11:38:02.834] -> 17:37:35 PINGREQ
````

**Comment by larsonmpdx** [comment]
I was able to reproduce my problem, see my updated pull request #754 

the res_init() call discussed above is important for its own sake but is not important to reproduce this bug, it happens with or without it.

steps to reproduce are:
-run the loop sample on a 32 bit arm device with a cell phone internet connection
-after the first packet is sent and acknowledged, disconnect internet with "nmcli radio wwan off"
-after 7 minutes, reconnect internet ("nmcli radio wwan on")
-after another 1-2 minutes, during the reconnection process, the azure iot library will start emitting error messages very quickly, I think during every dowork() call, and won't recover from this state


here is a console log:
````
# ./azure_loop_demo
Creating IoTHub handle
        ** iothub connected **
created packet with 4222 bytes
logging to azure
Confirmation callback received for message 1 with result IOTHUB_CLIENT_CONFIRMATION_OK




[disconnect internet here]




created packet with 4223 bytes
logging to azure
created packet with 4224 bytes
logging to azure
created packet with 4225 bytes
logging to azure
created packet with 4226 bytes
logging to azure
created packet with 4227 bytes
logging to azure
created packet with 4228 bytes
not logging to azure, queue full
created packet with 4229 bytes
not logging to azure, queue full
Confirmation callback received for message 2 with result IOTHUB_CLIENT_CONFIRMATION_MESSAGE_TIMEOUT
Error: Time:Mon Dec 10 16:39:56 2018 File:/azure-iot-sdk-c/c-utility/adapters/socketio_berkeley.c Func:lookup_address_and_initiate_socket_connection Line:283 Failure: getaddrinfo failure -3.
Error: Time:Mon Dec 10 16:39:56 2018 File:/azure-iot-sdk-c/c-utility/adapters/socketio_berkeley.c Func:socketio_open Line:766 lookup_address_and_connect_socket failed
Info: Closing tlsio from a state other than TLSIO_STATE_EXT_OPEN or TLSIO_STATE_EXT_ERROR
Error: Time:Mon Dec 10 16:39:56 2018 File:/azure-iot-sdk-c/c-utility/adapters/tlsio_openssl.c Func:on_underlying_io_open_complete Line:760 Invalid tlsio_state. Expected state is TLSIO_STATE_OPENING_UNDERLYING_IO.
Error: Time:Mon Dec 10 16:39:56 2018 File:/azure-iot-sdk-c/c-utility/adapters/tlsio_openssl.c Func:tlsio_openssl_open Line:1258 Failed opening the underlying I/O.
Error: Time:Mon Dec 10 16:39:56 2018 File:/azure-iot-sdk-c/umqtt/src/mqtt_client.c Func:mqtt_client_connect Line:1001 Error: io_open failed
Error: Time:Mon Dec 10 16:39:56 2018 File:/azure-iot-sdk-c/iothub_client/src/iothubtransport_mqtt_common.c Func:SendMqttConnectMsg Line:2141 failure connecting to address [address].
Confirmation callback received for message 3 with result IOTHUB_CLIENT_CONFIRMATION_MESSAGE_TIMEOUT
Error: Time:Mon Dec 10 16:40:32 2018 File:/azure-iot-sdk-c/c-utility/adapters/socketio_berkeley.c Func:lookup_address_and_initiate_socket_connection Line:283 Failure: getaddrinfo failure -3.
Error: Time:Mon Dec 10 16:40:32 2018 File:/azure-iot-sdk-c/c-utility/adapters/socketio_berkeley.c Func:socketio_open Line:766 lookup_address_and_connect_socket failed
Info: Closing tlsio from a state other than TLSIO_STATE_EXT_OPEN or TLSIO_STATE_EXT_ERROR
Error: Time:Mon Dec 10 16:40:32 2018 File:/azure-iot-sdk-c/c-utility/adapters/tlsio_openssl.c Func:on_underlying_io_open_complete Line:760 Invalid tlsio_state. Expected state is TLSIO_STATE_OPENING_UNDERLYING_IO.
Error: Time:Mon Dec 10 16:40:32 2018 File:/azure-iot-sdk-c/c-utility/adapters/tlsio_openssl.c Func:tlsio_openssl_open Line:1258 Failed opening the underlying I/O.
Error: Time:Mon Dec 10 16:40:32 2018 File:/azure-iot-sdk-c/umqtt/src/mqtt_client.c Func:mqtt_client_connect Line:1001 Error: io_open failed
Error: Time:Mon Dec 10 16:40:32 2018 File:/azure-iot-sdk-c/iothub_client/src/iothubtransport_mqtt_common.c Func:SendMqttConnectMsg Line:2141 failure connecting to address [address].
Confirmation callback received for message 4 with result IOTHUB_CLIENT_CONFIRMATION_MESSAGE_TIMEOUT
Confirmation callback received for message 5 with result IOTHUB_CLIENT_CONFIRMATION_MESSAGE_TIMEOUT
Error: Time:Mon Dec 10 16:41:08 2018 File:/azure-iot-sdk-c/c-utility/adapters/socketio_berkeley.c Func:lookup_address_and_initiate_socket_connection Line:283 Failure: getaddrinfo failure -3.
Error: Time:Mon Dec 10 16:41:08 2018 File:/azure-iot-sdk-c/c-utility/adapters/socketio_berkeley.c Func:socketio_open Line:766 lookup_address_and_connect_socket failed
Info: Closing tlsio from a state other than TLSIO_STATE_EXT_OPEN or TLSIO_STATE_EXT_ERROR
Error: Time:Mon Dec 10 16:41:08 2018 File:/azure-iot-sdk-c/c-utility/adapters/tlsio_openssl.c Func:on_underlying_io_open_complete Line:760 Invalid tlsio_state. Expected state is TLSIO_STATE_OPENING_UNDERLYING_IO.
Error: Time:Mon Dec 10 16:41:08 2018 File:/azure-iot-sdk-c/c-utility/adapters/tlsio_openssl.c Func:tlsio_openssl_open Line:1258 Failed opening the underlying I/O.
Error: Time:Mon Dec 10 16:41:08 2018 File:/azure-iot-sdk-c/umqtt/src/mqtt_client.c Func:mqtt_client_connect Line:1001 Error: io_open failed
Error: Time:Mon Dec 10 16:41:08 2018 File:/azure-iot-sdk-c/iothub_client/src/iothubtransport_mqtt_common.c Func:SendMqttConnectMsg Line:2141 failure connecting to address [address].
Confirmation callback received for message 6 with result IOTHUB_CLIENT_CONFIRMATION_MESSAGE_TIMEOUT
Error: Time:Mon Dec 10 16:41:44 2018 File:/azure-iot-sdk-c/c-utility/adapters/socketio_berkeley.c Func:lookup_address_and_initiate_socket_connection Line:283 Failure: getaddrinfo failure -3.
Error: Time:Mon Dec 10 16:41:44 2018 File:/azure-iot-sdk-c/c-utility/adapters/socketio_berkeley.c Func:socketio_open Line:766 lookup_address_and_connect_socket failed
Info: Closing tlsio from a state other than TLSIO_STATE_EXT_OPEN or TLSIO_STATE_EXT_ERROR
Error: Time:Mon Dec 10 16:41:44 2018 File:/azure-iot-sdk-c/c-utility/adapters/tlsio_openssl.c Func:on_underlying_io_open_complete Line:760 Invalid tlsio_state. Expected state is TLSIO_STATE_OPENING_UNDERLYING_IO.
Error: Time:Mon Dec 10 16:41:44 2018 File:/azure-iot-sdk-c/c-utility/adapters/tlsio_openssl.c Func:tlsio_openssl_open Line:1258 Failed opening the underlying I/O.
Error: Time:Mon Dec 10 16:41:44 2018 File:/azure-iot-sdk-c/umqtt/src/mqtt_client.c Func:mqtt_client_connect Line:1001 Error: io_open failed
Error: Time:Mon Dec 10 16:41:44 2018 File:/azure-iot-sdk-c/iothub_client/src/iothubtransport_mqtt_common.c Func:SendMqttConnectMsg Line:2141 failure connecting to address [address].
Error: Time:Mon Dec 10 16:42:20 2018 File:/azure-iot-sdk-c/c-utility/adapters/socketio_berkeley.c Func:lookup_address_and_initiate_socket_connection Line:283 Failure: getaddrinfo failure -3.
Error: Time:Mon Dec 10 16:42:20 2018 File:/azure-iot-sdk-c/c-utility/adapters/socketio_berkeley.c Func:socketio_open Line:766 lookup_address_and_connect_socket failed
Info: Closing tlsio from a state other than TLSIO_STATE_EXT_OPEN or TLSIO_STATE_EXT_ERROR
Error: Time:Mon Dec 10 16:42:20 2018 File:/azure-iot-sdk-c/c-utility/adapters/tlsio_openssl.c Func:on_underlying_io_open_complete Line:760 Invalid tlsio_state. Expected state is TLSIO_STATE_OPENING_UNDERLYING_IO.
Error: Time:Mon Dec 10 16:42:20 2018 File:/azure-iot-sdk-c/c-utility/adapters/tlsio_openssl.c Func:tlsio_openssl_open Line:1258 Failed opening the underlying I/O.
Error: Time:Mon Dec 10 16:42:20 2018 File:/azure-iot-sdk-c/umqtt/src/mqtt_client.c Func:mqtt_client_connect Line:1001 Error: io_open failed
Error: Time:Mon Dec 10 16:42:20 2018 File:/azure-iot-sdk-c/iothub_client/src/iothubtransport_mqtt_common.c Func:SendMqttConnectMsg Line:2141 failure connecting to address [address].
Error: Time:Mon Dec 10 16:42:56 2018 File:/azure-iot-sdk-c/c-utility/adapters/socketio_berkeley.c Func:lookup_address_and_initiate_socket_connection Line:283 Failure: getaddrinfo failure -3.
Error: Time:Mon Dec 10 16:42:56 2018 File:/azure-iot-sdk-c/c-utility/adapters/socketio_berkeley.c Func:socketio_open Line:766 lookup_address_and_connect_socket failed
Info: Closing tlsio from a state other than TLSIO_STATE_EXT_OPEN or TLSIO_STATE_EXT_ERROR
Error: Time:Mon Dec 10 16:42:56 2018 File:/azure-iot-sdk-c/c-utility/adapters/tlsio_openssl.c Func:on_underlying_io_open_complete Line:760 Invalid tlsio_state. Expected state is TLSIO_STATE_OPENING_UNDERLYING_IO.
Error: Time:Mon Dec 10 16:42:56 2018 File:/azure-iot-sdk-c/c-utility/adapters/tlsio_openssl.c Func:tlsio_openssl_open Line:1258 Failed opening the underlying I/O.
Error: Time:Mon Dec 10 16:42:56 2018 File:/azure-iot-sdk-c/umqtt/src/mqtt_client.c Func:mqtt_client_connect Line:1001 Error: io_open failed
Error: Time:Mon Dec 10 16:42:56 2018 File:/azure-iot-sdk-c/iothub_client/src/iothubtransport_mqtt_common.c Func:SendMqttConnectMsg Line:2141 failure connecting to address [address].
Error: Time:Mon Dec 10 16:43:32 2018 File:/azure-iot-sdk-c/c-utility/adapters/socketio_berkeley.c Func:lookup_address_and_initiate_socket_connection Line:283 Failure: getaddrinfo failure -3.
Error: Time:Mon Dec 10 16:43:32 2018 File:/azure-iot-sdk-c/c-utility/adapters/socketio_berkeley.c Func:socketio_open Line:766 lookup_address_and_connect_socket failed
Info: Closing tlsio from a state other than TLSIO_STATE_EXT_OPEN or TLSIO_STATE_EXT_ERROR
Error: Time:Mon Dec 10 16:43:32 2018 File:/azure-iot-sdk-c/c-utility/adapters/tlsio_openssl.c Func:on_underlying_io_open_complete Line:760 Invalid tlsio_state. Expected state is TLSIO_STATE_OPENING_UNDERLYING_IO.
Error: Time:Mon Dec 10 16:43:32 2018 File:/azure-iot-sdk-c/c-utility/adapters/tlsio_openssl.c Func:tlsio_openssl_open Line:1258 Failed opening the underlying I/O.
Error: Time:Mon Dec 10 16:43:32 2018 File:/azure-iot-sdk-c/umqtt/src/mqtt_client.c Func:mqtt_client_connect Line:1001 Error: io_open failed
Error: Time:Mon Dec 10 16:43:32 2018 File:/azure-iot-sdk-c/iothub_client/src/iothubtransport_mqtt_common.c Func:SendMqttConnectMsg Line:2141 failure connecting to address [address].
created packet with 4230 bytes
logging to azure




[after 7 minutes, reconnect internet]




Error: Time:Mon Dec 10 16:44:29 2018 File:/azure-iot-sdk-c/c-utility/adapters/tlsio_openssl.c Func:tlsio_openssl_send Line:1351 Invalid tlsio_state. Expected state is TLSIO_STATE_OPEN.
Error: Time:Mon Dec 10 16:44:29 2018 File:/azure-iot-sdk-c/umqtt/src/mqtt_client.c Func:sendPacketItem Line:371 1352: Failure sending control packet data
created packet with 4231 bytes
logging to azure
created packet with 4232 bytes
logging to azure
Error: Time:Mon Dec 10 16:45:00 2018 File:/azure-iot-sdk-c/iothub_client/src/iothubtransport_mqtt_common.c Func:InitializeConnection Line:2220 mqtt_client timed out waiting for CONNACK
Error: Time:Mon Dec 10 16:45:00 2018 File:/azure-iot-sdk-c/iothub_client/src/iothubtransport_mqtt_common.c Func:InitializeConnection Line:2220 mqtt_client timed out waiting for CONNACK
Error: Time:Mon Dec 10 16:45:00 2018 File:/azure-iot-sdk-c/iothub_client/src/iothubtransport_mqtt_common.c Func:InitializeConnection Line:2220 mqtt_client timed out waiting for CONNACK
Error: Time:Mon Dec 10 16:45:00 2018 File:/azure-iot-sdk-c/iothub_client/src/iothubtransport_mqtt_common.c Func:InitializeConnection Line:2220 mqtt_client timed out waiting for CONNACK
Error: Time:Mon Dec 10 16:45:00 2018 File:/azure-iot-sdk-c/iothub_client/src/iothubtransport_mqtt_common.c Func:InitializeConnection Line:2220 mqtt_client timed out waiting for CONNACK
Error: Time:Mon Dec 10 16:45:00 2018 File:/azure-iot-sdk-c/iothub_client/src/iothubtransport_mqtt_common.c Func:InitializeConnection Line:2220 mqtt_client timed out waiting for CONNACK
Error: Time:Mon Dec 10 16:45:01 2018 File:/azure-iot-sdk-c/iothub_client/src/iothubtransport_mqtt_common.c Func:InitializeConnection Line:2220 mqtt_client timed out waiting for CONNACK
Error: Time:Mon Dec 10 16:45:01 2018 File:/azure-iot-sdk-c/iothub_client/src/iothubtransport_mqtt_common.c Func:InitializeConnection Line:2220 mqtt_client timed out waiting for CONNACK
Error: Time:Mon Dec 10 16:45:01 2018 File:/azure-iot-sdk-c/iothub_client/src/iothubtransport_mqtt_common.c Func:InitializeConnection Line:2220 mqtt_client timed out waiting for CONNACK
Error: Time:Mon Dec 10 16:45:01 2018 File:/azure-iot-sdk-c/iothub_client/src/iothubtransport_mqtt_common.c Func:InitializeConnection Line:2220 mqtt_client timed out waiting for CONNACK
Error: Time:Mon Dec 10 16:45:01 2018 File:/azure-iot-sdk-c/iothub_client/src/iothubtransport_mqtt_common.c Func:InitializeConnection Line:2220 mqtt_client timed 

[this message continues forever and quickly fills up the logs]
````

**Comment by larsonmpdx** [comment]
I am able to reproduce this in an ubuntu 16.04 **32 bit** desktop VM running on vmware, using the 2018-11-20 release compiled from source, and just turning the network connection to the vm off and then back on after 7 minutes.  I was not able to reproduce on a 64-bit VM using a similar setup (2018-11-20 from apt-get instead of from source).

Please let me know if there is additional logging I can turn on or statements I can insert to gather more information.

**Comment by ewertons** [comment]
Hi @larsonmpdx ,
we will try your sample here.
In the meantime, could you try this? Set the retry policy on your sample to be IOTHUB_CLIENT_RETRY_INTERVAL.
Right after IoTHubDeviceClient_LL_CreateFromConnectionString, add this:

```c
(void)IoTHubDeviceClient_LL_SetRetryPolicy(device_ll_handle, IOTHUB_CLIENT_RETRY_INTERVAL, 0);
```

More info can be found in our documentation about [retry policies](https://github.com/Azure/azure-iot-sdk-c/blob/master/doc/connection_and_messaging_reliability.md#connection-retry-policies).

**Comment by larsonmpdx** [comment]
I added that line and it has the same result except the logs are longer as it retries more often when it's disconnected and then the "CONNACK" error spam starts more quickly after reconnecting the internet

**Comment by ewertons** [comment]
I see, thanks for trying that.
I will ask our field engineer to contact you to get some more details.

Have you tried using AMQP instead of MQTT for protocol?
If you can try that, please also enable [verbose logs](https://github.com/Azure/azure-iot-sdk-c/pull/754/files#diff-bb352a1a08fd67cc950d8b756b67c148R243).

**Comment by larsonmpdx** [comment]
Thanks, I had OPTION_LOG_TRACE=true set in my application before I created the standalone demo but I didn't turn it on for the demo.  Are there any additional logging options I can set?

I haven't tried aqmp, is there any advantage over mqtt? What do you recommend? I can try it for debugging if that might help.

Here is a sample with OPTION_LOG_TRACE=true.  the added lines are all of the lines with "->" or "<-".  In my full application I have previously seen different combinations of "-> DISCONNECT" and "-> PINGREQ", the only interesting additional messages after internet is disconnected, with the same result.

````
Creating IoTHub handle
-> 13:28:13 CONNECT | VER: 4 | KEEPALIVE: 240 | FLAGS: 128 | USERNAME: [address]/[device]/?api-version=2017-11-08-preview&DeviceClientType=iothubclient%2f1.2.11%20(native%3b%20Linux%3b%20i686) | CLEAN: 0
<- 13:28:13 CONNACK | SESSION_PRESENT: true | RETURN_CODE: 0x0
	** iothub connected **
created packet with 4222 bytes
logging to azure
-> 13:28:32 PUBLISH | IS_DUP: false | RETAIN: 0 | QOS: DELIVER_AT_LEAST_ONCE | TOPIC_NAME: devices/[device]/messages/events/ | PACKET_ID: 2 | PAYLOAD_LEN: 4222
<- 13:28:33 PUBACK | PACKET_ID: 2
Confirmation callback received for message 1 with result IOTHUB_CLIENT_CONFIRMATION_OK


created packet with 4223 bytes
logging to azure
-> 13:28:53 PUBLISH | IS_DUP: false | RETAIN: 0 | QOS: DELIVER_AT_LEAST_ONCE | TOPIC_NAME: devices/[device]/messages/events/ | PACKET_ID: 3 | PAYLOAD_LEN: 4223
created packet with 4224 bytes
logging to azure
-> 13:29:14 PUBLISH | IS_DUP: false | RETAIN: 0 | QOS: DELIVER_AT_LEAST_ONCE | TOPIC_NAME: devices/[device]/messages/events/ | PACKET_ID: 4 | PAYLOAD_LEN: 4224
created packet with 4225 bytes
logging to azure
-> 13:29:34 PUBLISH | IS_DUP: false | RETAIN: 0 | QOS: DELIVER_AT_LEAST_ONCE | TOPIC_NAME: devices/[device]/messages/events/ | PACKET_ID: 5 | PAYLOAD_LEN: 4225
-> 13:29:54 PUBLISH | IS_DUP: false | RETAIN: 0 | QOS: DELIVER_AT_LEAST_ONCE | TOPIC_NAME: devices/[device]/messages/events/ | PACKET_ID: 3 | PAYLOAD_LEN: 4223
created packet with 4226 bytes
logging to azure
-> 13:29:54 PUBLISH | IS_DUP: false | RETAIN: 0 | QOS: DELIVER_AT_LEAST_ONCE | TOPIC_NAME: devices/[device]/messages/events/ | PACKET_ID: 6 | PAYLOAD_LEN: 4226
-> 13:30:14 PUBLISH | IS_DUP: false | RETAIN: 0 | QOS: DELIVER_AT_LEAST_ONCE | TOPIC_NAME: devices/[device]/messages/events/ | PACKET_ID: 4 | PAYLOAD_LEN: 4224
created packet with 4227 bytes
logging to azure
-> 13:30:14 PUBLISH | IS_DUP: false | RETAIN: 0 | QOS: DELIVER_AT_LEAST_ONCE | TOPIC_NAME: devices/[device]/messages/events/ | PACKET_ID: 7 | PAYLOAD_LEN: 4227
created packet with 4228 bytes
warning: max iothub outstanding messages reached, logging next message offline
not logging to azure, queue full
-> 13:30:35 PUBLISH | IS_DUP: false | RETAIN: 0 | QOS: DELIVER_AT_LEAST_ONCE | TOPIC_NAME: devices/[device]/messages/events/ | PACKET_ID: 5 | PAYLOAD_LEN: 4225
created packet with 4229 bytes
warning: max iothub outstanding messages reached, logging next message offline
not logging to azure, queue full
Confirmation callback received for message 2 with result IOTHUB_CLIENT_CONFIRMATION_MESSAGE_TIMEOUT
-> 13:30:55 DISCONNECT
Error: Time:Wed Dec 12 13:30:57 2018 File:/home/user/azure-iot-sdk-c/c-utility/adapters/socketio_berkeley.c Func:lookup_address_and_initiate_socket_connection Line:283 Failure: getaddrinfo failure -3.
Error: Time:Wed Dec 12 13:30:57 2018 File:/home/user/azure-iot-sdk-c/c-utility/adapters/socketio_berkeley.c Func:socketio_open Line:766 lookup_address_and_connect_socket failed
Info: Closing tlsio from a state other than TLSIO_STATE_EXT_OPEN or TLSIO_STATE_EXT_ERROR
Error: Time:Wed Dec 12 13:30:57 2018 File:/home/user/azure-iot-sdk-c/c-utility/adapters/tlsio_openssl.c Func:on_underlying_io_open_complete Line:760 Invalid tlsio_state. Expected state is TLSIO_STATE_OPENING_UNDERLYING_IO.
Error: Time:Wed Dec 12 13:30:57 2018 File:/home/user/azure-iot-sdk-c/c-utility/adapters/tlsio_openssl.c Func:tlsio_openssl_open Line:1258 Failed opening the underlying I/O.
Error: Time:Wed Dec 12 13:30:57 2018 File:/home/user/azure-iot-sdk-c/umqtt/src/mqtt_client.c Func:mqtt_client_connect Line:1001 Error: io_open failed
Error: Time:Wed Dec 12 13:30:57 2018 File:/home/user/azure-iot-sdk-c/iothub_client/src/iothubtransport_mqtt_common.c Func:SendMqttConnectMsg Line:2141 failure connecting to address [address].
Error: Time:Wed Dec 12 13:30:58 2018 File:/home/user/azure-iot-sdk-c/c-utility/adapters/socketio_berkeley.c Func:lookup_address_and_initiate_socket_connection Line:283 Failure: getaddrinfo failure -3.
Error: Time:Wed Dec 12 13:30:58 2018 File:/home/user/azure-iot-sdk-c/c-utility/adapters/socketio_berkeley.c Func:socketio_open Line:766 lookup_address_and_connect_socket failed
Info: Closing tlsio from a state other than TLSIO_STATE_EXT_OPEN or TLSIO_STATE_EXT_ERROR
Error: Time:Wed Dec 12 13:30:58 2018 File:/home/user/azure-iot-sdk-c/c-utility/adapters/tlsio_openssl.c Func:on_underlying_io_open_complete Line:760 Invalid tlsio_state. Expected state is TLSIO_STATE_OPENING_UNDERLYING_IO.
Error: Time:Wed Dec 12 13:30:58 2018 File:/home/user/azure-iot-sdk-c/c-utility/adapters/tlsio_openssl.c Func:tlsio_openssl_open Line:1258 Failed opening the underlying I/O.
Error: Time:Wed Dec 12 13:30:58 2018 File:/home/user/azure-iot-sdk-c/umqtt/src/mqtt_client.c Func:mqtt_client_connect Line:1001 Error: io_open failed
Error: Time:Wed Dec 12 13:30:58 2018 File:/home/user/azure-iot-sdk-c/iothub_client/src/iothubtransport_mqtt_common.c Func:SendMqttConnectMsg Line:2141 failure connecting to address [address].
Error: Time:Wed Dec 12 13:31:00 2018 File:/home/user/azure-iot-sdk-c/c-utility/adapters/socketio_berkeley.c Func:lookup_address_and_initiate_socket_connection Line:283 Failure: getaddrinfo failure -3.
Error: Time:Wed Dec 12 13:31:00 2018 File:/home/user/azure-iot-sdk-c/c-utility/adapters/socketio_berkeley.c Func:socketio_open Line:766 lookup_address_and_connect_socket failed
Info: Closing tlsio from a state other than TLSIO_STATE_EXT_OPEN or TLSIO_STATE_EXT_ERROR
Error: Time:Wed Dec 12 13:31:00 2018 File:/home/user/azure-iot-sdk-c/c-utility/adapters/tlsio_openssl.c Func:on_underlying_io_open_complete Line:760 Invalid tlsio_state. Expected state is TLSIO_STATE_OPENING_UNDERLYING_IO.
Error: Time:Wed Dec 12 13:31:00 2018 File:/home/user/azure-iot-sdk-c/c-utility/adapters/tlsio_openssl.c Func:tlsio_openssl_open Line:1258 Failed opening the underlying I/O.
Error: Time:Wed Dec 12 13:31:00 2018 File:/home/user/azure-iot-sdk-c/umqtt/src/mqtt_client.c Func:mqtt_client_connect Line:1001 Error: io_open failed
Error: Time:Wed Dec 12 13:31:00 2018 File:/home/user/azure-iot-sdk-c/iothub_client/src/iothubtransport_mqtt_common.c Func:SendMqttConnectMsg Line:2141 failure connecting to address [address].
Error: Time:Wed Dec 12 13:31:04 2018 File:/home/user/azure-iot-sdk-c/c-utility/adapters/socketio_berkeley.c Func:lookup_address_and_initiate_socket_connection Line:283 Failure: getaddrinfo failure -3.
Error: Time:Wed Dec 12 13:31:04 2018 File:/home/user/azure-iot-sdk-c/c-utility/adapters/socketio_berkeley.c Func:socketio_open Line:766 lookup_address_and_connect_socket failed
Info: Closing tlsio from a state other than TLSIO_STATE_EXT_OPEN or TLSIO_STATE_EXT_ERROR
Error: Time:Wed Dec 12 13:31:04 2018 File:/home/user/azure-iot-sdk-c/c-utility/adapters/tlsio_openssl.c Func:on_underlying_io_open_complete Line:760 Invalid tlsio_state. Expected state is TLSIO_STATE_OPENING_UNDERLYING_IO.
Error: Time:Wed Dec 12 13:31:04 2018 File:/home/user/azure-iot-sdk-c/c-utility/adapters/tlsio_openssl.c Func:tlsio_openssl_open Line:1258 Failed opening the underlying I/O.
Error: Time:Wed Dec 12 13:31:04 2018 File:/home/user/azure-iot-sdk-c/umqtt/src/mqtt_client.c Func:mqtt_client_connect Line:1001 Error: io_open failed
Error: Time:Wed Dec 12 13:31:04 2018 File:/home/user/azure-iot-sdk-c/iothub_client/src/iothubtransport_mqtt_common.c Func:SendMqttConnectMsg Line:2141 failure connecting to address [address].
Error: Time:Wed Dec 12 13:31:12 2018 File:/home/user/azure-iot-sdk-c/c-utility/adapters/socketio_berkeley.c Func:lookup_address_and_initiate_socket_connection Line:283 Failure: getaddrinfo failure -3.
Error: Time:Wed Dec 12 13:31:12 2018 File:/home/user/azure-iot-sdk-c/c-utility/adapters/socketio_berkeley.c Func:socketio_open Line:766 lookup_address_and_connect_socket failed
Info: Closing tlsio from a state other than TLSIO_STATE_EXT_OPEN or TLSIO_STATE_EXT_ERROR
Error: Time:Wed Dec 12 13:31:12 2018 File:/home/user/azure-iot-sdk-c/c-utility/adapters/tlsio_openssl.c Func:on_underlying_io_open_complete Line:760 Invalid tlsio_state. Expected state is TLSIO_STATE_OPENING_UNDERLYING_IO.
Error: Time:Wed Dec 12 13:31:12 2018 File:/home/user/azure-iot-sdk-c/c-utility/adapters/tlsio_openssl.c Func:tlsio_openssl_open Line:1258 Failed opening the underlying I/O.
Error: Time:Wed Dec 12 13:31:12 2018 File:/home/user/azure-iot-sdk-c/umqtt/src/mqtt_client.c Func:mqtt_client_connect Line:1001 Error: io_open failed
Error: Time:Wed Dec 12 13:31:12 2018 File:/home/user/azure-iot-sdk-c/iothub_client/src/iothubtransport_mqtt_common.c Func:SendMqttConnectMsg Line:2141 failure connecting to address [address].
Confirmation callback received for message 3 with result IOTHUB_CLIENT_CONFIRMATION_MESSAGE_TIMEOUT
created packet with 4230 bytes
logging to azure
Error: Time:Wed Dec 12 13:31:28 2018 File:/home/user/azure-iot-sdk-c/c-utility/adapters/socketio_berkeley.c Func:lookup_address_and_initiate_socket_connection Line:283 Failure: getaddrinfo failure -3.
Error: Time:Wed Dec 12 13:31:28 2018 File:/home/user/azure-iot-sdk-c/c-utility/adapters/socketio_berkeley.c Func:socketio_open Line:766 lookup_address_and_connect_socket failed
Info: Closing tlsio from a state other than TLSIO_STATE_EXT_OPEN or TLSIO_STATE_EXT_ERROR
Error: Time:Wed Dec 12 13:31:28 2018 File:/home/user/azure-iot-sdk-c/c-utility/adapters/tlsio_openssl.c Func:on_underlying_io_open_complete Line:760 Invalid tlsio_state. Expected state is TLSIO_STATE_OPENING_UNDERLYING_IO.
Error: Time:Wed Dec 12 13:31:28 2018 File:/home/user/azure-iot-sdk-c/c-utility/adapters/tlsio_openssl.c Func:tlsio_openssl_open Line:1258 Failed opening the underlying I/O.
Error: Time:Wed Dec 12 13:31:28 2018 File:/home/user/azure-iot-sdk-c/umqtt/src/mqtt_client.c Func:mqtt_client_connect Line:1001 Error: io_open failed
Error: Time:Wed Dec 12 13:31:28 2018 File:/home/user/azure-iot-sdk-c/iothub_client/src/iothubtransport_mqtt_common.c Func:SendMqttConnectMsg Line:2141 failure connecting to address [address].
Confirmation callback received for message 4 with result IOTHUB_CLIENT_CONFIRMATION_MESSAGE_TIMEOUT
created packet with 4231 bytes
logging to azure
Confirmation callback received for message 5 with result IOTHUB_CLIENT_CONFIRMATION_MESSAGE_TIMEOUT
created packet with 4232 bytes
logging to azure
Error: Time:Wed Dec 12 13:32:01 2018 File:/home/user/azure-iot-sdk-c/c-utility/adapters/socketio_berkeley.c Func:lookup_address_and_initiate_socket_connection Line:283 Failure: getaddrinfo failure -3.
Error: Time:Wed Dec 12 13:32:01 2018 File:/home/user/azure-iot-sdk-c/c-utility/adapters/socketio_berkeley.c Func:socketio_open Line:766 lookup_address_and_connect_socket failed
Info: Closing tlsio from a state other than TLSIO_STATE_EXT_OPEN or TLSIO_STATE_EXT_ERROR
Error: Time:Wed Dec 12 13:32:01 2018 File:/home/user/azure-iot-sdk-c/c-utility/adapters/tlsio_openssl.c Func:on_underlying_io_open_complete Line:760 Invalid tlsio_state. Expected state is TLSIO_STATE_OPENING_UNDERLYING_IO.
Error: Time:Wed Dec 12 13:32:01 2018 File:/home/user/azure-iot-sdk-c/c-utility/adapters/tlsio_openssl.c Func:tlsio_openssl_open Line:1258 Failed opening the underlying I/O.
Error: Time:Wed Dec 12 13:32:01 2018 File:/home/user/azure-iot-sdk-c/umqtt/src/mqtt_client.c Func:mqtt_client_connect Line:1001 Error: io_open failed
Error: Time:Wed Dec 12 13:32:01 2018 File:/home/user/azure-iot-sdk-c/iothub_client/src/iothubtransport_mqtt_common.c Func:SendMqttConnectMsg Line:2141 failure connecting to address [address].
Confirmation callback received for message 6 with result IOTHUB_CLIENT_CONFIRMATION_MESSAGE_TIMEOUT
created packet with 4233 bytes
logging to azure
created packet with 4234 bytes
logging to azure
created packet with 4235 bytes
warning: max iothub outstanding messages reached, logging next message offline
not logging to azure, queue full
Error: Time:Wed Dec 12 13:33:06 2018 File:/home/user/azure-iot-sdk-c/c-utility/adapters/socketio_berkeley.c Func:lookup_address_and_initiate_socket_connection Line:283 Failure: getaddrinfo failure -3.
Error: Time:Wed Dec 12 13:33:06 2018 File:/home/user/azure-iot-sdk-c/c-utility/adapters/socketio_berkeley.c Func:socketio_open Line:766 lookup_address_and_connect_socket failed
Info: Closing tlsio from a state other than TLSIO_STATE_EXT_OPEN or TLSIO_STATE_EXT_ERROR
Error: Time:Wed Dec 12 13:33:06 2018 File:/home/user/azure-iot-sdk-c/c-utility/adapters/tlsio_openssl.c Func:on_underlying_io_open_complete Line:760 Invalid tlsio_state. Expected state is TLSIO_STATE_OPENING_UNDERLYING_IO.
Error: Time:Wed Dec 12 13:33:06 2018 File:/home/user/azure-iot-sdk-c/c-utility/adapters/tlsio_openssl.c Func:tlsio_openssl_open Line:1258 Failed opening the underlying I/O.
Error: Time:Wed Dec 12 13:33:06 2018 File:/home/user/azure-iot-sdk-c/umqtt/src/mqtt_client.c Func:mqtt_client_connect Line:1001 Error: io_open failed
Error: Time:Wed Dec 12 13:33:06 2018 File:/home/user/azure-iot-sdk-c/iothub_client/src/iothubtransport_mqtt_common.c Func:SendMqttConnectMsg Line:2141 failure connecting to address [address].
created packet with 4236 bytes
warning: max iothub outstanding messages reached, logging next message offline
not logging to azure, queue full
created packet with 4237 bytes
warning: max iothub outstanding messages reached, logging next message offline
not logging to azure, queue full
created packet with 4238 bytes
warning: max iothub outstanding messages reached, logging next message offline
not logging to azure, queue full
created packet with 4239 bytes
warning: max iothub outstanding messages reached, logging next message offline
not logging to azure, queue full
created packet with 4240 bytes
warning: max iothub outstanding messages reached, logging next message offline
not logging to azure, queue full
created packet with 4241 bytes
warning: max iothub outstanding messages reached, logging next message offline
not logging to azure, queue full
Error: Time:Wed Dec 12 13:35:15 2018 File:/home/user/azure-iot-sdk-c/c-utility/adapters/socketio_berkeley.c Func:lookup_address_and_initiate_socket_connection Line:283 Failure: getaddrinfo failure -3.
Error: Time:Wed Dec 12 13:35:15 2018 File:/home/user/azure-iot-sdk-c/c-utility/adapters/socketio_berkeley.c Func:socketio_open Line:766 lookup_address_and_connect_socket failed
Info: Closing tlsio from a state other than TLSIO_STATE_EXT_OPEN or TLSIO_STATE_EXT_ERROR
Error: Time:Wed Dec 12 13:35:15 2018 File:/home/user/azure-iot-sdk-c/c-utility/adapters/tlsio_openssl.c Func:on_underlying_io_open_complete Line:760 Invalid tlsio_state. Expected state is TLSIO_STATE_OPENING_UNDERLYING_IO.
Error: Time:Wed Dec 12 13:35:15 2018 File:/home/user/azure-iot-sdk-c/c-utility/adapters/tlsio_openssl.c Func:tlsio_openssl_open Line:1258 Failed opening the underlying I/O.
Error: Time:Wed Dec 12 13:35:15 2018 File:/home/user/azure-iot-sdk-c/umqtt/src/mqtt_client.c Func:mqtt_client_connect Line:1001 Error: io_open failed
Error: Time:Wed Dec 12 13:35:15 2018 File:/home/user/azure-iot-sdk-c/iothub_client/src/iothubtransport_mqtt_common.c Func:SendMqttConnectMsg Line:2141 failure connecting to address [address].
created packet with 4242 bytes
warning: max iothub outstanding messages reached, logging next message offline
not logging to azure, queue full
created packet with 4243 bytes
warning: max iothub outstanding messages reached, logging next message offline
not logging to azure, queue full




created packet with 4244 bytes
warning: max iothub outstanding messages reached, logging next message offline
not logging to azure, queue full
created packet with 4245 bytes
warning: max iothub outstanding messages reached, logging next message offline
not logging to azure, queue full
created packet with 4246 bytes
warning: max iothub outstanding messages reached, logging next message offline
not logging to azure, queue full
created packet with 4247 bytes
warning: max iothub outstanding messages reached, logging next message offline
not logging to azure, queue full
created packet with 4248 bytes
warning: max iothub outstanding messages reached, logging next message offline
not logging to azure, queue full
created packet with 4249 bytes
warning: max iothub outstanding messages reached, logging next message offline
not logging to azure, queue full
created packet with 4250 bytes
warning: max iothub outstanding messages reached, logging next message offline
not logging to azure, queue full
created packet with 4251 bytes
warning: max iothub outstanding messages reached, logging next message offline
not logging to azure, queue full
created packet with 4252 bytes
warning: max iothub outstanding messages reached, logging next message offline
not logging to azure, queue full
created packet with 4253 bytes
warning: max iothub outstanding messages reached, logging next message offline
not logging to azure, queue full
created packet with 4254 bytes
warning: max iothub outstanding messages reached, logging next message offline
not logging to azure, queue full
Error: Time:Wed Dec 12 13:39:42 2018 File:/home/user/azure-iot-sdk-c/c-utility/adapters/tlsio_openssl.c Func:tlsio_openssl_send Line:1351 Invalid tlsio_state. Expected state is TLSIO_STATE_OPEN.
Error: Time:Wed Dec 12 13:39:42 2018 File:/home/user/azure-iot-sdk-c/umqtt/src/mqtt_client.c Func:sendPacketItem Line:371 1352: Failure sending control packet data
-> 13:39:42 PINGREQ
created packet with 4255 bytes
warning: max iothub outstanding messages reached, logging next message offline
not logging to azure, queue full
created packet with 4256 bytes
warning: max iothub outstanding messages reached, logging next message offline
not logging to azure, queue full
Error: Time:Wed Dec 12 13:40:13 2018 File:/home/user/azure-iot-sdk-c/iothub_client/src/iothubtransport_mqtt_common.c Func:InitializeConnection Line:2220 mqtt_client timed out waiting for CONNACK
Error: Time:Wed Dec 12 13:40:13 2018 File:/home/user/azure-iot-sdk-c/iothub_client/src/iothubtransport_mqtt_common.c Func:InitializeConnection Line:2220 mqtt_client timed out waiting for CONNACK
Error: Time:Wed Dec 12 13:40:13 2018 File:/home/user/azure-iot-sdk-c/iothub_client/src/iothubtransport_mqtt_common.c Func:InitializeConnection Line:2220 mqtt_client timed out waiting for CONNACK
Error: Time:Wed Dec 12 13:40:13 2018 File:/home/user/azure-iot-sdk-c/iothub_client/src/iothubtransport_mqtt_common.c Func:InitializeConnection Line:2220 mqtt_client timed out waiting for CONNACK
Error: Time:Wed Dec 12 13:40:13 2018 File:/home/user/azure-iot-sdk-c/iothub_client/src/iothubtransport_mqtt_common.c Func:InitializeConnection Line:2220 mqtt_client timed out waiting for CONNACK
Error: Time:Wed Dec 12 13:40:13 2018 File:/home/user/azure-iot-sdk-c/iothub_client/src/iothubtransport_mqtt_common.c Func:InitializeConnection Line:2220 mqtt_client timed out waiting for CONNACK
Error: Time:Wed Dec 12 13:40:13 2018 File:/home/user/azure-iot-sdk-c/iothub_client/src/iothubtransport_mqtt_common.c Func:InitializeConnection Line:2220 mqtt_client timed out waiting for CONNACK
````

**Comment by ewertons** [comment]
This log is good, thanks.

Sure, I understand your final implementation choice will be MQTT. 
The reason of the ask to run using AMQP is for testing only, so we can see if there is common connection issue, or if it's specific to the MQTT transport.

**Comment by larsonmpdx** [comment]
switching to amqp it can now reconnect without an error

what are the tradeoffs vs mqtt?  this page only lists increased resource use for amqp.  is packet size similar?

https://docs.microsoft.com/en-us/azure/iot-hub/iot-hub-devguide-protocols

**Comment by larsonmpdx** [comment]
We're unblocked for development as we're able to test all features without triggering the bug.  We'd like to use mqtt for our deployment and just need to produce a final binary for acceptance testing at this point.

**Comment by ewertons** [comment]
Just to mention, a missing information in the description relates to the way the device client is being authenticated.
The customer is using x509 certs for authentication.

**Comment by ewertons** [comment]
@larsonmpdx ,

getaddrinfo returning -3 really comes from the underlying system, leaving no recourse to the IoT Hub device client.
We could not verify why with AMQP transport you didn't see the reconnection errors. 

However, 

we ran some tests, and we were able to get a similar output to yours. 

The steps to reproduce were:
1. Start iothub device client (using MQTT, x509 auth);
> It connects and starts sending 1 telemetry per second;
2. After a few seconds, disable network interface: 
`ifconfig eth0 down`
3. After 2 minutes, transport detects network issue and starts trying to reconnect;
> At this point there are several "Failure: getaddrinfo failure -3" failures, indicating the name resolution in the system is faulty.
> In parallel, verified that name resolution fails by running "ping www.bing.com".

4. Re-enabled network interface;
`ifconfig eth0 up`
> Verified name resolution by running `ping www.bing.com`, it still fails!

So at this point the iothub device client is still trying to reconnect, but failing with getaddrinfo setting errno to -3. That is expected since the platform is not providing name resolution to the SDK.

Then, I proceed with these steps to attempt to fix the name resolution:

5. Release the previous address reservation:
`dhclient -r eth0`
6. Renew the address lease:
`dhclient eth0`
> Now verified that name resolution is work again, running `ping www.bing.com`.
7. Verified that iothub device client could now reconnect on its own.

Once the platform started providing name resolution correctly, the IoT hub device client could reconnect.

So, could you run this verification in your platform as well?

Note: used a Hyper-V VM running Debian 9.5.0 amd64.

**Comment by larsonmpdx** [comment]
getaddrinfo() errors are not related to the problem we're seeing, please see discussion up thread on this. 
 The problem is that, upon reconnection, the azure device sdk will enter a loop wherein it prints an error seemingly with every call to dowork(), and doesn't recover from this state.  here is the error you're looking for:

````
Error: Time:Wed Dec 12 13:40:13 2018 File:/home/user/azure-iot-sdk-c/iothub_client/src/iothubtransport_mqtt_common.c Func:InitializeConnection Line:2220 mqtt_client timed out waiting for CONNACK
Error: Time:Wed Dec 12 13:40:13 2018 File:/home/user/azure-iot-sdk-c/iothub_client/src/iothubtransport_mqtt_common.c Func:InitializeConnection Line:2220 mqtt_client timed out waiting for CONNACK
Error: Time:Wed Dec 12 13:40:13 2018 File:/home/user/azure-iot-sdk-c/iothub_client/src/iothubtransport_mqtt_common.c Func:InitializeConnection Line:2220 mqtt_client timed out waiting for CONNACK
Error: Time:Wed Dec 12 13:40:13 2018 File:/home/user/azure-iot-sdk-c/iothub_client/src/iothubtransport_mqtt_common.c Func:InitializeConnection Line:2220 mqtt_client timed out waiting for CONNACK
...
````

In my testing, the problem is isolated to 32-bit platforms, and I needed a particular sequence of packets (see my reproducing sample, about 4KB every 20 seconds) and reconnect delay (7 minutes) to trigger it.  I think you'll need to install a new 32-bit linux instance and run my sample to see it.

**Comment by jebrando** [comment]
@jebgarcia 

I'm looking at the verbose log you added and I'm noticing something.  On line 8 there is a PUBACK from the service.  This is a response to the PUBLISH message on line 7, this means that the device and the service are connected and the messages got received correctly.  After that there are many PUBLISH messages, but no PUBACK messages.  This means that the service is not receiving any more messages from the devices.  In this situation the device disconnects and attempts to reestablish network connections.  It looks like when the device attempts to reconnect it gets an error from getaddrinfo (-3 EAI_AGAIN - The name server returned a temporary failure indication.  Try again later).  This is pointing to something's up with the DNS.  I'm going to try to repro this failure today and see if there is anything the SDK can do to mitigate this failure.

Thanks
Jelani

**Comment by jebrando** [comment]
@jebgarcia I do not have Virtual Box on my setup, and I tried to install it, but I'm going to need to reinstall and it will take a min so I tried it on docker instead.  Here are the steps I tried for the repro:

1. Get a docker 32bit image of debian on my wifi connected laptop
```Shell
docker pull 32bit/debian
docker run -it -v <path_to_folder containing_sdk>:/home 32bit/debian:latest bash
apt-get update && apt-get upgrade
apt-get install -y curl build-essential pkg-config libcurl4-openssl-dev git cmake libssl-dev uuid-dev valgrind
```
2. Copy my sample over the iothub telemetry sample and add a valid connection string in the variable
[iothub_ll_telemetry_sample.zip](https://github.com/Azure/azure-iot-sdk-c/files/2738150/iothub_ll_telemetry_sample.zip)


3. Build sample on the docker container.  This sample will sends a telemetry message every 30 seconds
```Shell
cd /home/azure-iot-sdk-c
mkdir cmake
cd cmake
cmake ..
cd iothub_client/samples/iothub_ll_telemetry_sample/
make -j
```
3. Run sample and after the first PUBACK turn off the laptop wifi
```Shell
./iothub_ll_telemetry_sample
```
4. Let the sample detect the disconnect and sit for a few min
5. Turn back on the wif
6. The Sample worked as expected

If you can try this with my sample and let me know if you get the same result.

I have seen an error in the MQTT library where a PING and CONNECT gets sent out while the device is not connected.  I'll try to get a fix in on this, but it shouldn't be a blocker for this.

Let me know what you come up with.

**Comment by larsonmpdx** [comment]
I have only been able to reproduce this with the demonstration code in this pr on a 32 bit os.  the demo code sends some ~4KB packets on about a 20 second schedule, it seemed necessary to do something like that rather than the smaller messages on a different schedule sent by the existing example code.

https://github.com/Azure/azure-iot-sdk-c/pull/754

If you try reproducing it with an unedited iothub_ll_telemetry_sample I'm not sure it will show you the problem

**Comment by jebrando** [comment]
@larsonmpdx I've been looking at the error more and I'm pretty sure that the issue that you're seeing has been fixed in the latest SDK (this would explain why I can't reproduce it).  I believe that the fix was checked in sometime in Dec.  Upgrade to the latest and it should be fixed for you.

**Comment by Patrik-Berglund** [comment]
@jebrando With the latest SDK, are you refering to the 2018-12-13 release?

We are currently using that release and have just found logs flooded with never ending:

```
Error: Time:Mon Jan 14 19:57:00 2019 File:/home/patrik/Patrik-Berglund/esp32-azure-iot/azure-iot-sdk-c/iothub_client/src/iothubtransport_mqtt_common.c Func:InitializeConnection Line:2229 mqtt_client timed out waiting for CONNACK
Error: Time:Mon Jan 14 19:57:00 2019 File:/home/patrik/Patrik-Berglund/esp32-azure-iot/azure-iot-sdk-c/iothub_client/src/iothubtransport_mqtt_common.c Func:InitializeConnection Line:2229 mqtt_client timed out waiting for CONNACK
```

**Comment by jebrando** [comment]
The I updated the latest release of the SDK to reconnect the on 32bit systems.  Let me know if you see anything different.

**Comment by larsonmpdx** [comment]
thanks, I can test this

were you able to reproduce the original problem, and then see that this change fixes it in your reproduction?

**Comment by AndWass** [comment]
@jebrando @Patrik-Berglund I have also noticed that the issue appears to be fixed. Running on a GSM-connection that used to give me issues every now and then. Will continue testing as well.

**Comment by Patrik-Berglund** [comment]
@AndWass @jebrando Finished the first test session without any problems. Started the next one now with intermittent connectivity. 

When can we see these fixes in a release? Wan't to be able to push to production and roll out a new SW release after next week if possible. If we don't run in to any problems...

**Comment by massand** [comment]
This fix is in master now. We are releasing this fix next week. Thanks for the follow up!

**Comment by AndWass** [comment]
@massand Thanks for releasing next week!
@Patrik-Berglund Could you keep me posted on the intermittent connectivity tests (just post results here or something)?

**Comment by jebrando** [comment]
@AndWass Thanks for the update.  If you find anything else, let us know.

**Comment by Patrik-Berglund** [comment]
@AndWass 

Hi, we have completed the second week of tests and have seen on issue. Trying to dig deeper in to the problem.. Will probably open a new issue for tracking.

The issue is related to the following two events,

Had a total failure in our DEV environment IoT Hub during last night (started 20190201 20:54 UTC). Lost all 18 devices at the same time for 3 hours. The devices is located at 8 different places using different ISP:s. This generated a hefty 100MB logs from the 10 devices in the lab..

Had a similar event at 2019-01-28 23:55 to 2019-01-29 00:00 but it didn't affect all devices (although multiple locations).

Have checked https://azure.microsoft.com/sv-se/status/history/ to see if any major outage in the IoT Hub service tool place.

@AndWass  located in Sweden?

**Comment by AndWass** [comment]
@Patrik-Berglund

Interesting, we haven’t deployed this version to many devices yet. Did the devices reconnect again on their own? I have previously had an issue where a fault in the hub would cause a device to not reconnect, but that was with AMQP. 

Will keep an eye out for the new issue. 

Yes Sweden indeed, you as well?

**Comment by Patrik-Berglund** [comment]
@AndWass 

The logs during that time span are filled with this
1. The connection is opened (status=0 reason=6)
2. We send a heartbeat message up on opening
3. The connection is closed (status=1 reason=4)

And that repeats roughly every third second for three hours... and this happened to all 18 devices, at the same time.

Have update the code now so we don't send the heartbeat directly, avoiding a flood of messages next time this might happen.

Yes, Sweden... it's a small country, we have actually worked together in a project, although different areas. :smiley:

```
[0;32mI (93792) IOT HUB: Connection Status Callback: status=0 reason=6[0m
-> 21:45:29 SUBSCRIBE | PACKET_ID: 99 | TOPIC_NAME: devices/240ac412c060/messages/devicebound/# | QOS: 1 | TOPIC_NAME: $iothub/twin/res/# | QOS: 0 | TOPIC_NAME: $iothub/methods/POST/# | QOS: 0

[0;32mI (93902) Main: Sending heartbeat..[0m
[0;32mI (93902) IOT HUB: Sending HeartbeatTelemetry[0m
[0;32mI (93912) IOT HUB: IoTHubClient accepted the message for delivery[0m
-> 21:45:30 PUBLISH | IS_DUP: false | RETAIN: 0 | QOS: DELIVER_AT_LEAST_ONCE | TOPIC_NAME: devices/240ac412c060/messages/events/message-type=HeartbeatTelemetry&message-version=1 | PACKET_ID: 100 | PAYLOAD_LEN: 413

<- 21:45:30 SUBACK | PACKET_ID: 99 | RETURN_CODE: 1 | RETURN_CODE: 0 | RETURN_CODE: 0

-> 21:45:30 PUBLISH | IS_DUP: false | RETAIN: 0 | QOS: DELIVER_AT_MOST_ONCE | TOPIC_NAME: $iothub/twin/GET/?$rid=101

[0;32mI (94472) IOT HUB: Connection Status Callback: status=1 reason=4[0m
[0;33mW (94472) IOT HUB: IOTHUB_CLIENT_CONNECTION_NO_NETWORK[0m
Error: Time:Fri Feb  1 21:45:30 2019 File:/home/buildagent/vsts-agent/_work/1/s/components/azure-iot/pal/src/socketio_berkeley.c Func:socketio_send Line:834 Failure: socket state is not opened.

-> 21:45:32 CONNECT | VER: 4 | KEEPALIVE: 240 | FLAGS: 192 | USERNAME: cgrind-dev-iothub01.azure-devices.net/240ac412c060/?api-version=2017-11-08-preview&DeviceClientType=iothubclient%2f1.2.13%20(openssl_compact) | PWD: XXXX | CLEAN: 0

<- 21:45:32 CONNACK | SESSION_PRESENT: true | RETURN_CODE: 0x0

[0;32mI (96362) IOT HUB: Connection Status Callback: status=0 reason=6[0m
-> 21:45:32 SUBSCRIBE | PACKET_ID: 102 | TOPIC_NAME: devices/240ac412c060/messages/devicebound/# | QOS: 1 | TOPIC_NAME: $iothub/twin/res/# | QOS: 0 | TOPIC_NAME: $iothub/methods/POST/# | QOS: 0

<- 21:45:32 SUBACK | PACKET_ID: 102 | RETURN_CODE: 1 | RETURN_CODE: 0 | RETURN_CODE: 0

[0;32mI (96422) Main: Sending heartbeat..[0m
[0;32mI (96422) IOT HUB: Sending HeartbeatTelemetry[0m
-> 21:45:32 PUBLISH | IS_DUP: false | RETAIN: 0 | QOS: DELIVER_AT_MOST_ONCE | TOPIC_NAME: $iothub/twin/GET/?$rid=103

[0;32mI (96442) IOT HUB: IoTHubClient accepted the message for delivery[0m
-> 21:45:32 PUBLISH | IS_DUP: false | RETAIN: 0 | QOS: DELIVER_AT_LEAST_ONCE | TOPIC_NAME: devices/240ac412c060/messages/events/message-type=HeartbeatTelemetry&message-version=1 | PACKET_ID: 104 | PAYLOAD_LEN: 413

[0;32mI (96992) IOT HUB: Connection Status Callback: status=1 reason=4[0m
[0;33mW (96992) IOT HUB: IOTHUB_CLIENT_CONNECTION_NO_NETWORK[0m
```
